hazelcast/src/main/java/com/hazelcast/internal/memory/impl/DirectMemoryBits.java

```diff
@@ -263,77 +371,103 @@ public static long readLong(long address, boolean bigEndian) {
+    }
+
+    public static long readLong(MemoryAccessor memoryAccessor, Object base, long offset, boolean bigEndian) {
         if (bigEndian) {
-            return readLongB(base, offset);
+            return readLongB(memoryAccessor, base, offset);
         } else {
-            return readLongL(base, offset);
+            return readLongL(memoryAccessor, base, offset);
         }
     }
 
     public static long readLongB(Object base, long offset) {
-        long byte7 = (long) UNSAFE.getByte(base, offset) << 56;
-        long byte6 = (long) (UNSAFE.getByte(base, offset + 1) & 0xFF) << 48;
-        long byte5 = (long) (UNSAFE.getByte(base, offset + 2) & 0xFF) << 40;
-        long byte4 = (long) (UNSAFE.getByte(base, offset + 3) & 0xFF) << 32;
-        long byte3 = (long) (UNSAFE.getByte(base, offset + 4) & 0xFF) << 24;
-        long byte2 = (long) (UNSAFE.getByte(base, offset + 5) & 0xFF) << 16;
-        long byte1 = (long) (UNSAFE.getByte(base, offset + 6) & 0xFF) << 8;
-        long byte0 = (long) (UNSAFE.getByte(base, offset + 7) & 0xFF);
-        return byte7 + byte6 + byte5 + byte4 + byte3 + byte2 + byte1 + byte0;
+        return readLongB(STANDARD_MEMORY_ACCESSOR, base, offset);
     }
 
     public static long readLongL(Object base, long offset) {
-        long byte7 = (long) (UNSAFE.getByte(base, offset) & 0xFF);
-        long byte6 = (long) (UNSAFE.getByte(base, offset + 1) & 0xFF) << 8;
-        long byte5 = (long) (UNSAFE.getByte(base, offset + 2) & 0xFF) << 16;
-        long byte4 = (long) (UNSAFE.getByte(base, offset + 3) & 0xFF) << 24;
-        long byte3 = (long) (UNSAFE.getByte(base, offset + 4) & 0xFF) << 32;
-        long byte2 = (long) (UNSAFE.getByte(base, offset + 5) & 0xFF) << 40;
-        long byte1 = (long) (UNSAFE.getByte(base, offset + 6) & 0xFF) << 48;
-        long byte0 = (long) (UNSAFE.getByte(base, offset + 7) & 0xFF) << 56;
+        return readLongL(STANDARD_MEMORY_ACCESSOR, base, offset);
+    }
+
+    public static long readLongB(MemoryAccessor memoryAccessor, Object base, long offset) {
+        long byte7 = (long) memoryAccessor.getByte(base, offset) << 56;
+        long byte6 = (long) (memoryAccessor.getByte(base, offset + 1) & 0xFF) << 48;
+        long byte5 = (long) (memoryAccessor.getByte(base, offset + 2) & 0xFF) << 40;
+        long byte4 = (long) (memoryAccessor.getByte(base, offset + 3) & 0xFF) << 32;
+        long byte3 = (long) (memoryAccessor.getByte(base, offset + 4) & 0xFF) << 24;
+        long byte2 = (long) (memoryAccessor.getByte(base, offset + 5) & 0xFF) << 16;
+        long byte1 = (long) (memoryAccessor.getByte(base, offset + 6) & 0xFF) << 8;
+        long byte0 = (long) (memoryAccessor.getByte(base, offset + 7) & 0xFF);
+        return byte7 + byte6 + byte5 + byte4 + byte3 + byte2 + byte1 + byte0;
+    }
+
+    public static long readLongL(MemoryAccessor memoryAccessor, Object base, long offset) {
+        long byte7 = (long) (memoryAccessor.getByte(base, offset) & 0xFF);
```

Comment:
As a consequence of the current design, this method will perform eight validation checks, one for each invocation of `getByte`. Maybe we should rethink this.